python/lsst/daf/butler/_formatter.py

```diff
@@ -51,12 +63,1058 @@
+          component and derive.
+
+        Disassembled Composite:
+
+        * The file to read here is the component itself. Formatter only knows
+          about this one component file. Should be no component specified
+          in the ``read`` call but the `FileDescriptor` will know which
+          component this is.
+        * A derived component. The file to read is a component but not the
+          specified component. The caching needs the component from which
+          it's derived.
+
+        Raises
+        ------
+        FormatterNotImplementedError
+            Raised if no implementations were found that could read this
+            resource.
+        """
+        # If the file to read is a ZIP file with a fragment requesting
+        # a file within the ZIP file, it is no longer possible to use the
+        # direct read from URI option.
+        uri = self.file_descriptor.location.uri
+        if uri.getExtension() == ".zip" and uri.fragment and uri.fragment.startswith("zip-path="):
+            # Preference to use URIs with "file.zip#zip-path=thing.json"
+            # rather than "file.zip#thing.json"
+            _, path_in_zip = uri.fragment.split("=")
+
+            # Open the Zip file using ResourcePath.
+            with uri.open("rb") as fd:
+                with zipfile.ZipFile(fd) as zf:  # type: ignore
+                    if self.can_read_from_stream:
+                        zip_fd = zf.open(path_in_zip)
+                        try:
+                            in_memory_dataset = self.read_from_stream(zip_fd, component)
+                        except Exception:
+                            zip_fd.close()
+                            raise
+                        return in_memory_dataset
+
+                    # For now for both URI and local file options we retrieve
+                    # the bytes to a temporary local and use that.
+                    _, suffix = os.path.splitext(path_in_zip)
+                    with ResourcePath.temporary_uri(suffix=suffix) as tmp_uri:
+                        tmp_uri.write(zf.read(path_in_zip))
+
+                        if self.can_read_from_local_file:
+                            return self.read_from_local_file(tmp_uri, component)
+                        if self.can_read_from_uri:
+                            return self.read_from_uri(tmp_uri, component)
+
+            raise FormatterNotImplementedError(
+                f"Formatter {self.name()} could not read the file using any method."
+            )
+
+        # First see if the formatter can support direct remote read from
+        # a URI. We allow FormatterNotImplementedError to be raised by
+        # the formatter in case the formatter wishes to defer to the
+        # local file implementation which will trigger a cache write.
+        if self.can_read_from_uri:
+            with contextlib.suppress(FormatterNotImplementedError):
+                return self.read_directly_from_possibly_cached_uri(
```

Comment:
I wonder if we should trap explicitly the no component / no parameters case and not call the read_from_uri here but defer it to the later situation where we have downloaded and possibly cached the file and then can pass the local URI to read_from_uri. That might be better than having the formatter author decide for themselves whether they might want to use the cached file instead.

Reply:
I was thinking of the non-component read case since that one will need the full file anyhow (if there are no parameters). If a formatter author believes that it's always better to read from the local file then they should only implement `read_from_local_file`. What we don't want is for an author to implement `read_from_uri` by transferring the whole file locally and then opening it -- that would break caching. I think I need to add that to the documentation.

As currently implemented, if someone asks for a component and `read_from_uri` can handle a component read without downloading the whole file, then the whole file won't be cached.

Reply:
I have changed the code so that read_from_uri can be called later after downloading to the cache if `read_from_uri` raised not implemented error. What do you think of always downloading the file to local storage (as we do now) if there is no component and not parameters specified?

Reply:
Modified to do the download and cache if there are no components or parameters specified.